### PR TITLE
book-store-impl

### PR DESCRIPTION
**Pull request by SkillfulDev**

### CODE REVIEW

File: src/main/java/ua/chernonog/onlinebookstore/repository/impl/BookRepositoryImpl.java [hunk 1]
```diff
@@ -1,19 +1,51 @@
...
+            transaction = entityManager.getTransaction();
+            transaction.begin();
+            entityManager.persist(book);
+            transaction.commit();
+            return book;
+        } catch (RuntimeException ex) {
+            if (transaction != null && transaction.isActive()) {
+                transaction.rollback();
+            }
+            throw new EntitySaveException("Can`t save book " + book + " to DB");
```
Comment (ReamFOX) [inline src/main/java/ua/chernonog/onlinebookstore/repository/impl/BookRepositoryImpl.java]:
```suggestion
            throw new EntitySaveException("Can`t save book " + book + " to DB", e);
```

File: src/main/java/ua/chernonog/onlinebookstore/exception/EntitySaveException.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package ua.chernonog.onlinebookstore.exception;
+
+public class EntitySaveException extends RuntimeException {
+    public EntitySaveException(String message) {
+        super(message);
```
Comment (ReamFOX) [inline src/main/java/ua/chernonog/onlinebookstore/exception/EntitySaveException.java]:
```suggestion
    public EntitySaveException(String message, Throwable e) {
        super(message, e);
```
I think would be better pass e to this exception, if you use it only in try-catch of repository

**Review by ReamFOX — APPROVED** [review]
GJ, just a minor comment

File: src/main/java/ua/chernonog/onlinebookstore/repository/impl/BookRepositoryImpl.java (hunk 1 above)
Comment (SkillfulDev) [reply src/main/java/ua/chernonog/onlinebookstore/repository/impl/BookRepositoryImpl.java]: good idea, but unfortunately I`ve already changed my implementation with data JPA.

**Review by ReamFOX — APPROVED** [review]
gj

File: src/main/java/ua/chernonog/onlinebookstore/exception/EntitySaveException.java (hunk 2 above)
Comment (andrii-hoienko) [reply src/main/java/ua/chernonog/onlinebookstore/exception/EntitySaveException.java]: agree

File: src/main/java/ua/chernonog/onlinebookstore/repository/impl/BookRepositoryImpl.java (hunk 1 above)
Comment (andrii-hoienko) [reply src/main/java/ua/chernonog/onlinebookstore/repository/impl/BookRepositoryImpl.java]: agree

File: src/main/java/ua/chernonog/onlinebookstore/repository/impl/BookRepositoryImpl.java [hunk 3]
```diff
@@ -1,19 +1,51 @@
...
+                    .getResultList();
+        }
+    }
+
+    @Override
+    public Optional<Book> findById(Long id) {
+        try (EntityManager entityManager = entityManagerFactory.createEntityManager()) {
+            Book book = entityManager.find(Book.class, id);
+            return Optional.ofNullable(book);
+        }
```
Comment (andrii-hoienko) [inline src/main/java/ua/chernonog/onlinebookstore/repository/impl/BookRepositoryImpl.java]: Where is catch blocks?

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Good job, but some improvements required

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Please resolve all comments

File: src/main/java/ua/chernonog/onlinebookstore/service/BookServiceImpl.java [hunk 4]
```diff
@@ -1,18 +1,41 @@
...
 
     @Override
-    public List<Book> findAll() {
-        return null;
+    public BookDto findBookById(Long id) {
+        Optional<Book> bookFromDB = bookRepository.findById(id);
+        if (bookFromDB.isPresent()) {
+            return bookMapper.toDto(bookFromDB.get());
+        }
+        throw new EntityNotFoundException("Can`t find book with id " + id + " in DB");
```
Comment (yevheniisynytsia) [inline src/main/java/ua/chernonog/onlinebookstore/service/BookServiceImpl.java]: can be replaced by `orElseThrow`

**Review by yevheniisynytsia — APPROVED** [review]
Please pay attention to the comment I've left

**Pull request merged**